aws_signing_helper/serve.go

```diff
@@ -0,0 +1,270 @@
+	}
+
+	// Handles requests to /latest/meta-data/iam/security-credentials/
+	getRoleNameHandler := func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "GET" {
+			w.WriteHeader(http.StatusMethodNotAllowed)
+			return
+		}
+
+		err := checkValidToken(w, r)
+		if err != nil {
+			return
+		}
+
+		io.WriteString(w, roleName) // nosemgrep
+	}
+
+	// Handles GET requests to /latest/meta-data/iam/security-credentials/<ROLE_NAME>
+	getCredentialsHandler := func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "GET" {
+			w.WriteHeader(http.StatusMethodNotAllowed)
+			return
+		}
+
+		err := checkValidToken(w, r)
+		if err != nil {
+			return
+		}
+
+		var nextRefreshTime = cred.Expiration.Add(-RefreshTime)
+		if time.Until(nextRefreshTime) < RefreshTime {
+			credentialProcessOutput, _ := GenerateCredentials(opts)
+			cred.AccessKeyId = credentialProcessOutput.AccessKeyId
+			cred.SecretAccessKey = credentialProcessOutput.SecretAccessKey
+			cred.Token = credentialProcessOutput.SessionToken
+			cred.Expiration, _ = time.Parse(time.RFC3339, credentialProcessOutput.Expiration)
+			err := json.NewEncoder(w).Encode(cred)
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				io.WriteString(w, "failed to encode credentials")
+				return
+			}
+		} else {
+			err := json.NewEncoder(w).Encode(cred)
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				io.WriteString(w, "failed to encode credentials")
+				return
+			}
+		}
+	}
+
+	return putTokenHandler, getRoleNameHandler, getCredentialsHandler
+}
+
+func Serve(port int, credentialsOptions CredentialsOpts) {
+	var refreshableCred = RefreshableCred{}
+
+	roleArn, err := arn.Parse(credentialsOptions.RoleArn)
+	if err != nil {
+		log.Println("invalid role ARN")
+		syscall.Exit(1)
+	}
+
+	credentialProcessOutput, _ := GenerateCredentials(&credentialsOptions)
+	refreshableCred.AccessKeyId = credentialProcessOutput.AccessKeyId
+	refreshableCred.SecretAccessKey = credentialProcessOutput.SecretAccessKey
+	refreshableCred.Token = credentialProcessOutput.SessionToken
+	refreshableCred.Expiration, _ = time.Parse(time.RFC3339, credentialProcessOutput.Expiration)
+	endpoint := &Endpoint{PortNum: port, TmpCred: refreshableCred}
+	endpoint.Server = &http.Server{}
+	roleResourceParts := strings.Split(roleArn.Resource, "/")
+	roleName := roleResourceParts[len(roleResourceParts)-1] // Find role name without path
+	putTokenHandler, getRoleNameHandler, getCredentialsHandler := AllIssuesHandlers(&endpoint.TmpCred, roleName, &credentialsOptions)
+
+	http.HandleFunc(TOKEN_RESOURCE_PATH, putTokenHandler)
+	http.HandleFunc(SECURITY_CREDENTIALS_RESOURCE_PATH, getRoleNameHandler)
+	http.HandleFunc(SECURITY_CREDENTIALS_RESOURCE_PATH+roleName, getCredentialsHandler)
+
+	// Background thread that cleans up expired tokens
+	ticker := time.NewTicker(5 * time.Second)
+	go func() {
+		for range ticker.C {
+			curTime := time.Now()
+			mutex.Lock()
```

Comment:
I think `sync.Map` will let you implement this logic w/out managing your own `Mutex`